src/Aspire.Hosting/ParameterResourceBuilderExtensions.cs

```diff
@@ -27,6 +27,43 @@ public static IResourceBuilder<ParameterResource> AddParameter(this IDistributed
         return builder.AddParameter(name, parameterDefault => GetParameterValue(builder.Configuration, name, parameterDefault), secret: secret);
     }
 
+    /// <summary>
+    /// Adds a parameter resource to the application, providing a default value to be used as a fallback.
+    /// </summary>
+    /// <param name="builder">Distributed application builder</param>
+    /// <param name="name">Name of parameter resource</param>
+    /// <param name="defaultValue">A string value that is used if the name is not found in the configuration</param>
+    /// <param name="secret">Optional flag indicating whether the parameter should be regarded as secret.</param>
+    /// <returns>Resource builder for the parameter.</returns>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters",
+                                                     Justification = "third parameters are mutually exclusive.")]
+    public static IResourceBuilder<ParameterResource> AddParameter(this IDistributedApplicationBuilder builder, string name, string defaultValue, bool secret = false)
+    {
+        // An alternate implementation is to use some ConstantParameterDefault implementation that always returns the default value.
+        // However, doing this causes a "default": {} to be written to the manifest, which is not valid.
+        // And note that we ignore parameterDefault in the callback, because it can never be non-null, and we want our own default.
+        return builder.AddParameter(name, parameterDefault => builder.Configuration[$"Parameters:{name}"] ?? defaultValue, secret: secret);
+    }
+
+    /// <summary>
+    /// Adds a parameter resource to the application, providing a ParameterDefault to be used as a fallback.
+    /// </summary>
+    /// <param name="builder">Distributed application builder</param>
+    /// <param name="name">Name of parameter resource</param>
+    /// <param name="defaultValue">A ParameterDefault that is used if the name is not found in the configuration</param>
+    /// <param name="secret">Optional flag indicating whether the parameter should be regarded as secret.</param>
+    /// <returns>Resource builder for the parameter.</returns>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters",
+                                                     Justification = "third parameters are mutually exclusive.")]
+    public static IResourceBuilder<ParameterResource> AddParameter(this IDistributedApplicationBuilder builder, string name, ParameterDefault defaultValue, bool secret = false)
+    {
+        return builder.AddParameter(
+            name,
+            parameterDefault => GetParameterValue(builder.Configuration, name, parameterDefault),
```

Comment:
What should be the "interaction" with UserSecretsParameterDefault here?

When we generate a default password for things like PostgreSQL databases, we generate the password once, and then store it in User Secrets of the AppHost so it doesn't get generated on the next run. Should this do something similar?

https://github.com/dotnet/aspire/blob/6db273bbef306b1c62ad27f91816ea095a9df972/src/Aspire.Hosting/ParameterResourceBuilderExtensions.cs#L190-L201

Reply:
I see your point. Two options:
1. Have the user call ParameterResourceBuilderExtensions.CreateGeneratedParameter() themselves if they want that behavior. And then pass that to the new `AddParameter` overload.
2. Do it under the cover for them

And yes, 2. is probably better. And now I better understand why @DamianEdwards wanted different overloads for plain `ParameterDefault` vs `GenerateParameterDefault`, which makes this slightly cleaner (but is mostly indistinguishable for the user).

If that sounds good, I'll go ahead and do that.

Reply:
> Have the user call ParameterResourceBuilderExtensions.CreateGeneratedParameter() themselves if they want that behavior. And then pass that to the new AddParameter overload.

`ParameterResourceBuilderExtensions.CreateGeneratedParameter()` returns a `ParameterResource` (one that isn't directly attached to the App Model). So it can't be passed to the new AddParameter overload - since that takes a `ParameterDefault`.

I think the distinction between plain `ParameterDefault` vs `GenerateParameterDefault` isn't that obvious - especially if one overload does something different.

Do you (or @DamianEdwards or @davidfowl or @mitchdenny) have any thoughts on adding a `bool persistDefaultValue = false` parameter? Or potentially using a different method name for the method that persists the value in user secrets.

Reply:
Sorry, yes, my `ParameterResourceBuilderExtensions.CreateGeneratedParameter` suggestion made no sense. I meant something like:

```c#
public static IResourceBuilder<ParameterResource> AddParameter(this IDistributedApplicationBuilder builder, string name, GenerateParameterDefault defaultValue, bool secret = false)
{
    ParameterDefault paramDefault = defaultValue;
    if (builder.ExecutionContext.IsRunMode && builder.AppHostAssembly is not null)
    {
        paramDefault = new UserSecretsParameterDefault(builder.AppHostAssembly, builder.Environment.ApplicationName, name, paramDefault);
    }

    return builder.AddParameter(
        name,
        paramDefault,
        secret: secret);
}
```

Your `persistDefaultValue` suggestion is potentially interesting.
- Pros: avoids special casing GenerateParameterDefault. And there could be other ParameterDefault implementations that need persistence.
- Cons: a bit harder to use, if we assume that GenerateParameterDefault should almost always be persisted

Would like to hear what others thing as well.

Reply:
@eerhardt I ended up going with a variation:
- Single method that takes in a base `ParameterDefault`. No special case for `GenerateParameterDefault`.
- Instead, persistent is driven by a new `ParameterDefault.NeedsPersistence` property, which is false by default, and that `GenerateParameterDefault` sets to true. I kept `NeedsPersistence` internal, but we could expose (now or later)

PTAL

Reply:
I think this works. I like that we don't expose it until we need to. But it gives us a nice way of opening it up in the future.

Can you write a test for it?

Reply:
Thanks for the review. I added a simple test that checks that the `UserSecretsParameterDefault` wrapping only happens when it should.